Readme.md

```diff
@@ -0,0 +1,1432 @@
+
+```scala
+val votesByLang = votes groupBy { case (lang, _) => lang }
+val sumByLang = votesByLang map {
+  case (lang, counts) =>
+    val countsOnly = counts map { case (_, count) => count }
+    (lang, countsOnly.sum)
+}
+val orderedVotes = sumByLang.toSeq
+  .sortBy { case (_, count) => count }
+  .reverse
+```
+
+## Operations nesting and `for` comprehensions
+
+`for` comprehensions are a very nice syntactic sugar in Scala which allows expressing certain patterns in a very clean way. Of course, just like any other feature, it is crucial not to overuse it.
+
+As a rule of thumb, prefer `for` comprehensions to nested `flatMap`s and `map`s, for example, when you need to flatten several collections. `for` comprehensions are rewritten internally into a nested sequence of method calls, so be careful for using `for`s in performance-critical code.
+
+Be aware of implications of using `for` comprehensions for futures. See the section on concurrency for more on that.
+
+### Using `for` comprehensions for different types
+
+`for` comprehensions may be used with types which have `flatMap`, `map` and `withFilter` methods, and they do not guard you from using different types for each binding clause if it is allowed by typing:
+
+```scala
+val ints = Vector(1, 2, 4, 5)
+val m = Map(1 -> "a", 3 -> "c")
+
+for {
+  n <- ints
+  s <- m.get(n)
+} yield s
+```
+
+This compiles, even though `ints` is of type `Vector[Int]` and `m.get(n)` is of type `Option[String]`. This is compiled into
+
+```scala
+ints.flatMap(n => m.get(n))
+```
+
+`flatMap` on Scala collections allows its argument function to return anything convertible to an iterable, and `Option[T]` satisfies this requirement. This makes the code harder to understand and, in some subtle cases, may lead to unexpected behavior, in particular, when collections with uniqueness semantics of keys or values (like maps or sets) are used. Therefore, in most cases when you need to iterate over multiple collections in a nested fashion, convert them to a common type. Usually it makes sense to use `.iterator` for that:
+
+```scala
+val resultsIterator = for {
+  (interface, peers) <- signals.iterator
+  (peer, values) <- peers.iterator
+  value <- values.iterator
+} yield interface -> value
+
+val results = resultsIterator.toMap
+```
+
+## String interpolation
+
+Scala allows writing references to other variables in scope inside strings using a special syntax:
+
+```scala
+val n = 10
+println(s"The value is $n, for your information")
+```
+
+This syntax should be used instead of string formatting:
+
+```scala
+// bad
+println("The value is %s, for your information".format(n))
+```
+
+and instead of explicit concatenation:
+
+```scala
+// also bad
+println("The value is " + n + ", for your information")
+```
+
+Avoid using long expressions inside the string, though; if computation is long, store it in a lazy variable or a `def`:
```

Comment:
Just to define more precisely, what exactly the long expressions are,
I would permit only direct values or field access inside strings, e.g.:
``s"Hello, $name!"``,
``s"Hello, ${person.name}!`` and even
``s"Hello, ${person.fullName.firstName}!"``
are good, but something like
``s"Hello, ${userService.getPersonName(userId)}!"``
looks very obscure, IMO.

I believe, if some computation is required, there's no reason why it could not be extracted into a separate ``val``:
```scala
val personName = userService.getPersonName(userId)
s"Hello, $personName!"
```